### ISSUE TITLE
Trouble triggered and event load with new data from cloud

### ISSUE THREAD

**Issue by jdeltoft** [aleksanderwozniak/table_calendar#472]
I'm changing the events but calling setState() doesn't paint them.  I have to select a new date and then the dots for events show up over the calnedar days.  Is there a way to trigger a repaint of the calendar "events" programmatically without having to select a new day?

**Comment by aleksanderwozniak** [comment]
It sounds like you might be calling `setState()` before the events are actually updated, but it's hard to tell exactly without seeing the code. Are you using a `StreamBuilder` or `FutureBuilder` for this?

**Comment by jdeltoft** [comment]
Sure, I can try to add some example here.  I just was thinking maybe there is some obvious way to trigger a rebuild of the calendar.  So the example repo of the calendar I grabbled had "easy" constant data to show.  It would never not be available.  But what I want is to pull the latest list of events from the cloud, in the stateful widget that the calendar is shown in.  

Some vague examples of this exist on stackoverflow, but they appear to be using an earlier version I think of this library because the TableCalendar() has an "events:" field instead of "eventLoader:".  

https://stackoverflow.com/questions/60276303/flutter-table-calendar-showing-event-from-api-to-table-calendar
https://stackoverflow.com/questions/62715663/how-to-add-events-from-external-api-to-a-calender-in-flutter
https://stackoverflow.com/questions/64735251/adding-json-data-to-flutter-table-calendar-dynamically

The issue has several questions without a lot of clarity of a specific fix on stackoverflow.  I tried using ValueNotifier and ValueListenableBuilder as a way to try to "nudge" a rebuild of the calendar's event indicators once my data came in from my cloud api.  But no luck.  

If I simply click any other date but the initial starting date (today) it shows my events perfectly.  The issue I feel is that even a call to setState() doesn't trigger TableCalender() to force it's "eventLoader:" to run again.  I tried to see if Flutter Keys might be an answer but haven't found a solution there either.

Here are some code snippets of what I think are the important parts of my stateful calendar page:

```

  @override
  void initState() {
    _selectedDay = _focusedDay;
    getCloudData();
    _selectedDayReservations = ValueNotifier(_getReservationsForDay(_selectedDay!));
    _selectedDayTimes = ValueNotifier(_getSlotsForDay(_selectedDay!));
    _rebuildCalendar = ValueNotifier(0);
    super.initState();
  }

  void getCloudData() async {
    if (initCloudDataPulled == null || DateTime.now().difference(initCloudDataPulled!) > Duration(seconds: 2)) {
      initCloudDataPulled = DateTime.now();
      _parseReservationData(await cfGetAccountReservations());
      await cfGetSchedule(justDate.format(_selectedDay!));
      await cfGetDailyFill(justDate.format(_selectedDay!));
    }
  }


   .......
          ValueListenableBuilder(
            builder: (BuildContext context, int value, _) {
              return TableCalendar(
                firstDay: kFirstDay,
                lastDay: kLastDay,
                eventLoader: _getReservationsForDay,
                focusedDay: _focusedDay,
                availableCalendarFormats: {CalendarFormat.month: 'Month'},
                calendarFormat: _calendarFormat,
                selectedDayPredicate: (day) {
                  return isSameDay(_selectedDay, day);
                },
                onDaySelected: (selectedDay, focusedDay) {
                  if (!isSameDay(_selectedDay, selectedDay)) {
                    setState(() {
                      _selectedDay = selectedDay;
                      cfGetDailyFill(justDate.format(_selectedDay!));
                      _focusedDay = focusedDay;
                    });
                    _selectedDayReservations.value = _getReservationsForDay(selectedDay);
                    _selectedDayTimes.value = _getSlotsForDay(selectedDay);
                  }
                },
                onFormatChanged: (format) {
                  if (_calendarFormat != format) {
                    setState(() {
                      _calendarFormat = format;
                    });
                  }
                },
                onPageChanged: (focusedDay) {
                  // No need to call `setState()` here
                  _focusedDay = focusedDay;
                },
              );
            },
            valueListenable: _rebuildCalendar,
          ),
   .......


  List<Event> _getReservationsForDay(DateTime day) {
    print("got here day:$day");
    //return kReservations[day] ?? [];
    List<Event> retList = [];
    myReservations.forEach((date, eventList) {
      if (justDate.format(day) == justDate.format(date)) {
        retList.addAll(eventList);
      }
    });
    return retList;
  }

  void _parseReservationData(dynamic data) {
      print("reservation data: $data");
      var d = '2021-04-12 07:10:00';
      print(DateTime.parse(d));
      myReservations = LinkedHashMap<DateTime, List<Event>>( equals: isSameDay, hashCode: getHashCode
      ); //..addAll(_kEventSource);
      myReservations.addAll({DateTime.parse(d):[Event('Test 1')]});
      myReservations.addAll({DateTime.now():[Event('TEST 2')]});
      SchedulerBinding.instance!.addPostFrameCallback((_) {
        setState(() {
          print("rebuild after");
          _rebuildCalendar.value = 1;
        });
      });
  }
```

When I run this, I see all days of the month get called FIRST which makes sense.  This is the calendar using _getReservationsForDay() to get it's data for building.  THEN I see my _parseReservationData() finish it's call.  

In summary, my issue is I can't find any way to programmatically trigger a call to make TableCalendar() rebuild with _getReservationsForDay() again now that I have new data.

**Comment by jdeltoft** [comment]
Ah, I think I figured it out with StreamBuilder.  Something I should have really learned a while ago.  :D

**Comment by aleksanderwozniak** [comment]
Glad you got it to work now!

**Comment by zhanggenlove** [comment]
> 嗯，我想我已经用StreamBuilder解决了。我之前应该真正学到的东西。：D

Can you give me a simple demo? I have the same problem! I don’t know to render the calendar when I get event data from network apl
